### DISCUSSION TITLE
In the docker environment, after running for several hours, the connection cannot be made

### DISCUSSION THREAD

**Discussion by wangx024** [dani-garcia/vaultwarden#3588] [Q&A]
The error log is as follows
<html>
<body>

date | stream | content
-- | -- | --
2023-06-15T05:38:16.540242436Z | stdout | [2023-06-15 05:38:16.539][vaultwarden::api::core::two_factor][ERROR] Failed to get DB connection in send_incomplete_2fa_notifications()
2023-06-15T05:38:16.539731503Z | stdout | [2023-06-15 05:38:16.539][vaultwarden::db][ERROR] Timeout waiting for database connection
2023-06-15T05:37:49.504021646Z | stdout | [2023-06-15 05:37:49.502][rocket::server::_][WARN] No 503 catcher registered. Using Rocket default.
2023-06-15T05:37:49.503855548Z | stdout | [2023-06-15 05:37:49.502][vaultwarden::api::web::_][WARN] Request guard `DbConn` failed: ().
2023-06-15T05:37:49.503181277Z | stdout | [2023-06-15 05:37:49.502][vaultwarden::db][ERROR] Timeout waiting for database connection

</body>
</html>

**Comment by BlackDex** [comment]
Is the database still reacable? Are there any error messages on the database side?
That Vaultwarden can't reach the database isn't per definition an issue of Vaultwarden.

It might also be a docker issue, like docker updated, or a firewall issue, a lot of items which could cause an issue like this.

The only thing i can say in regards to what you posted here is, the database was lost, and Vaultwarden wasn't able to connect to it.

**Comment by krg3** [comment]
I can't offer any suggestions yet, but I wanted to say I have the same issue. My instance has been running for about four months. Yesterday it failed. My logs are the same as OP. 

[2023-06-16 09:13:41.391][vaultwarden::db][ERROR] Timeout waiting for database connection
[2023-06-16 09:13:41.391][vaultwarden::api::web::_][WARN] Request guard `DbConn` failed: ().
[2023-06-16 09:13:41.392][rocket::server::_][WARN] No 503 catcher registered. Using Rocket default.
[2023-06-16 09:14:23.518][vaultwarden::db][ERROR] Timeout waiting for database connection
[2023-06-16 09:14:23.518][vaultwarden::api::core::two_factor][ERROR] Failed to get DB connection in send_incomplete_2fa_notifications()
[2023-06-16 09:14:51.509][vaultwarden::db][ERROR] Timeout waiting for database connection
[2023-06-16 09:14:51.510][vaultwarden::api::web::_][WARN] Request guard `DbConn` failed: ().
[2023-06-16 09:14:51.510][rocket::server::_][WARN] No 503 catcher registered. Using Rocket default.
[2023-06-16 09:15:23.522][vaultwarden::db][ERROR] Timeout waiting for database connection
[2023-06-16 09:15:23.522][vaultwarden::api::core::two_factor][ERROR] Failed to get DB connection in send_incomplete_2fa_notifications()
[2023-06-16 09:15:45.949][rocket::server::_][WARN] Remote left: channel closed.
[2023-06-16 09:16:01.582][vaultwarden::db][ERROR] Timeout waiting for database connection
[2023-06-16 09:16:01.582][vaultwarden::api::web::_][WARN] Request guard `DbConn` failed: ().
[2023-06-16 09:16:01.582][rocket::server::_][WARN] No 503 catcher registered. Using Rocket default.
[2023-06-16 09:16:23.525][vaultwarden::db][ERROR] Timeout waiting for database connection
[2023-06-16 09:16:23.525][vaultwarden::api::core::two_factor][ERROR] Failed to get DB connection in send_incomplete_2fa_notifications()
[2023-06-16 09:17:11.668][vaultwarden::db][ERROR] Timeout waiting for database connection
[2023-06-16 09:17:11.668][vaultwarden::api::web::_][WARN] Request guard `DbConn` failed: ().
[2023-06-16 09:17:11.668][rocket::server::_][WARN] No 503 catcher registered. Using Rocket default.
[2023-06-16 09:17:23.527][vaultwarden::db][ERROR] Timeout waiting for database connection
[2023-06-16 09:17:23.528][vaultwarden::api::core::two_factor][ERROR] Failed to get DB connection in send_incomplete_2fa_notifications()

If I restart the container it runs for a few hours and fails again.

**Comment by BlackDex** [comment]
For both, i would suggest to see what has been updated in the past few days. Like docker, or an other container maybe?
As long as you are not using the `:testing` images of Vaultwarden, there is nothing changed, so it should be something else.

That Vaultwarden can't connect to the database is not instantly an issue with Vaultwarden, but could be somewhere else.

**Comment by krg3** [comment]
Isn't the DB internal to the image? I'm running only Vaultwarden and nginx proxy manager on this VM. No updates were run prior to the issue occurring. I did update the OS and docker images to try to fix the problem when it was discovered yesterday. Running Centos 9. Maybe @wangx024 is running Centos as well? Perhaps something in common there?

---
version: "2.1"
services:
  vaultwarden:
    image: vaultwarden/server:latest
    container_name: vaultwarden
    environment:
      - PUID=1000
      - PGID=1000
      - TZ=[redacted]
      - ADMIN_TOKEN=[redacted]
      - SIGNUPS_ALLOWED=false
      - SMTP_HOST=[redacted]
      - SMTP_FROM=[redacted]
      - SMTP_PORT=587
      - SMTP_SECURITY=starttls
      - SMTP_USERNAME=[redacted]
      - SMTP_PASSWORD=[redacted]
      - SMTP_AUTH_MECHANISM="Login"
      - LOG_FILE=/data/vaultwarden.log
    volumes:
      - /home/vaultwarden/data:/data
    networks:
      - vwd
    restart: unless-stopped
  npm:
    image: 'jc21/nginx-proxy-manager:latest'
    restart: unless-stopped
    ports:
      - '80:80'
      - '81:81'
      - '443:443'
    volumes:
      - ./npm:/data
      - ./letsencrypt:/etc/letsencrypt
    networks:
      - vwd
networks:
  vwd:
    external: true

**Comment by BlackDex** [comment]
If using sqlite, it's just a file.
If that is the case, then the next question would be, what type of filesystem is used. Is the Vaultwarden data on a shared storage? What kind of host is used? Are there backups created outside of the container, and does that in some way lock the file?

Because this kind of message is a bit strange.

Could you both provide the support string which you can generate via the admin interface? `/admin/diagnostics`

**Comment by krg3** [comment]
VM is xfs on top of LVM. Outside the VM we are running vshpere --> NFS --> NetApp FAS.

Backups are pulled daily. Veeam takes a snapshot and sends that off to alternate storage while the VM keeps humming. A quiesce is called during the snap creation. Doesn't seem likely to be related as the backups are scheduled at 00:00 and the logs on my machine show the DB going away at other times (09:13, 13:05). 

Support string below:

### Your environment (Generated via diagnostics page)
* Vaultwarden version: v1.28.1
* Web-vault version: v2023.3.0b
* OS/Arch: linux/x86_64
* Running within Docker: true (Base: Debian)
* Environment settings overridden: true
* Uses a reverse proxy: true
* IP Header check: true (X-Real-IP)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Browser/Server Time Check: true
* Server/NTP Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: SQLite
* Database version: 3.39.2
* Clients used: 
* Reverse proxy and version: 
* Other relevant information: 

### Config (Generated via diagnostics page)
<details><summary>Show Running Config</summary>

**Environment settings which are overridden:** SIGNUPS_ALLOWED, ADMIN_TOKEN, SMTP_HOST, SMTP_SECURITY, SMTP_PORT, SMTP_FROM, SMTP_USERNAME, SMTP_PASSWORD, SMTP_AUTH_MECHANISM


```json
{
  "_duo_akey": null,
  "_enable_duo": false,
  "_enable_email_2fa": true,
  "_enable_smtp": true,
  "_enable_yubico": false,
  "_icon_service_csp": "",
  "_icon_service_url": "",
  "_ip_header_enabled": true,
  "_smtp_img_src": "cid:",
  "admin_ratelimit_max_burst": 3,
  "admin_ratelimit_seconds": 300,
  "admin_session_lifetime": 20,
  "admin_token": "***",
  "allowed_iframe_ancestors": "",
  "attachments_folder": "data/attachments",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_conn_init": "",
  "database_max_conns": 10,
  "database_timeout": 30,
  "database_url": "***************",
  "db_connection_retries": 15,
  "disable_2fa_remember": false,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://************************",
  "domain_origin": "*****://************************",
  "domain_path": "",
  "domain_set": true,
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "email_attempts_limit": 3,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "emergency_access_allowed": true,
  "emergency_notification_reminder_schedule": "0 3 * * * *",
  "emergency_request_timeout_schedule": "0 7 * * * *",
  "enable_db_wal": true,
  "event_cleanup_schedule": "0 10 0 * * *",
  "events_days_retain": null,
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": null,
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "icon_redirect_code": 302,
  "icon_service": "internal",
  "incomplete_2fa_schedule": "30 * * * * *",
  "incomplete_2fa_time_limit": 3,
  "invitation_expiration_hours": 120,
  "invitation_org_name": "******************",
  "invitations_allowed": true,
  "ip_header": "X-Real-IP",
  "job_poll_interval_ms": 30000,
  "log_file": "/data/vaultwarden.log",
  "log_level": "Info",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "login_ratelimit_max_burst": 10,
  "login_ratelimit_seconds": 60,
  "org_attachment_limit": null,
  "org_creation_users": "",
  "org_events_enabled": false,
  "org_groups_enabled": false,
  "password_hints_allowed": true,
  "password_iterations": 100000,
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sendmail_command": null,
  "sends_allowed": false,
  "sends_folder": "data/sends",
  "show_password_hint": false,
  "signups_allowed": false,
  "signups_domains_whitelist": "",
  "signups_verify": false,
  "signups_verify_resend_limit": 6,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": "Login",
  "smtp_debug": false,
  "smtp_embed_images": true,
  "smtp_explicit_tls": null,
  "smtp_from": "************************",
  "smtp_from_name": "******************",
  "smtp_host": "******************",
  "smtp_password": "***",
  "smtp_port": 587,
  "smtp_security": "starttls",
  "smtp_ssl": null,
  "smtp_timeout": 5,
  "smtp_username": "************************",
  "templates_folder": "data/templates",
  "tmp_folder": "data/tmp",
  "trash_auto_delete_days": null,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_sendmail": false,
  "use_syslog": false,
  "user_attachment_limit": null,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "websocket_address": "0.0.0.0",
  "websocket_enabled": false,
  "websocket_port": 3012,
  "yubico_client_id": null,
  "yubico_secret_key": null,
  "yubico_server": null
}
```
</details>

**Comment by BlackDex** [comment]
Well, NFS has known issues with file locking, depending on which version of NFS. Also, if for some reason there is a small network hickup that could cause a strange behavior.

I would suggest to check if there were any updates on your host/hypervisor, NFS drivers/versions, the NetApp appliances, maybe even the network switches.

Because, if this behaviour just recently started, it probably isn't Vaultwarden, since that hasn't been updated since 3rd of April anymore.
